compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1384,9 +1384,11 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             }
 
             (ProjectionCandidate(i), ProjectionCandidate(j)) => {
-                // Arbitrarily pick the first candidate for backwards
+                // Arbitrarily pick the lower numbered candidate for backwards
                 // compatibility reasons. Don't let this affect inference.
-                i > j && !needs_infer
+                other.evaluation.must_apply_modulo_regions()
+                    && !needs_infer
+                    && (i < j || !victim.evaluation.must_apply_modulo_regions())
```

Comment:
Hmm I was worried when I read the PR description, but this is related narrowly tailored